docs/source/how_to_guides/cb_estimators.ipynb

```diff
@@ -0,0 +1,161 @@
+    "# Assumes examples have CBLabel typed labels.\n",
+    "def extract_label(examples: List[vw.Example]) -> Optional[Tuple[int, float, float]]:\n",
+    "    first_is_shared = len(examples) > 0 and examples[0].get_label().shared\n",
+    "    for i, example in enumerate(examples):\n",
+    "        if (label := example.get_label().label) is not None:\n",
+    "            _, cost, prob = label\n",
+    "            return (i - (1 if first_is_shared else 0), cost, prob)\n",
+    "    return None"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We'll use the following trivial input for this example. There are two actions, each identified by a single feature. We're using a StringIO so we can treat this as if we were reading it from a file with a {py:class}`vowpal_wabbit_next.TextFormatReader`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import io\n",
+    "\n",
+    "input = io.StringIO(\n",
+    "    \"\"\"shared | s\n",
+    "0:1:0.5 | a=0\n",
+    "| a=1\n",
+    "\n",
+    "shared | s\n",
+    "| a=0\n",
+    "1:0:0.5 | a=1\n",
+    "\n",
+    "shared | s\n",
+    "0:1:0.5 | a=0\n",
+    "| a=1\n",
+    "\n",
+    "shared | s\n",
+    "| a=0\n",
+    "1:0:0.5 | a=1\"\"\"\n",
+    ")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "See comments for an explanation of the process."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Estimate: -0.2625000001862645\n",
+      "Lower bound: -1.0\n",
+      "Upper bound: 0.3219763298424875\n"
+     ]
+    }
+   ],
+   "source": [
+    "workspace = vw.Workspace([\"--cb_explore_adf\"])\n",
+    "estimator = ips.Estimator()\n",
+    "interval = cressieread.Interval(empirical_r_bounds=True)\n",
+    "\n",
+    "estimates = []\n",
+    "lower = []\n",
+    "upper = []\n",
+    "\n",
+    "with vw.TextFormatReader(workspace, input) as reader:\n",
+    "    for event in reader:\n",
+    "        logged_label = extract_label(event)\n",
+    "\n",
+    "        # 1. Check if this event is labelled, if not skip it\n",
+    "        if logged_label is None:\n",
+    "            continue\n",
+    "\n",
+    "        # 2. Predict and learn on the event\n",
+    "        pmf = workspace.predict_then_learn_one(event)\n",
+    "\n",
+    "        # 3. Extract the logged cost and the probability of choosing it according to the logged policy\n",
+    "        logged_action_0_based, logged_cost, logged_prob = logged_label\n",
+    "\n",
+    "        # 4. Get the probability of choosing the logged action according to the target policy\n",
+    "        prediction_prob = next(x for i, x in pmf if i == logged_action_0_based)\n",
```

Comment:
[nit] dict(pmf)[logged_action_0_based] (not sure which one is more readable)

Reply:
Oh interesting, I didn't know you could init a dictionary like that